File: server/src/main/java/org/jetbrains/bsp/bazel/server/bloop/SourceSetRewriter.java
package org.jetbrains.bsp.bazel.server.bloop;

import io.vavr.collection.HashSet;
import io.vavr.collection.Set;
import java.nio.file.Path;
import java.nio.file.Paths;
import org.jetbrains.bsp.bazel.server.sync.model.SourceSet;

class SourceSetRewriter {
  private static final SourceSet EMPTY_SOURCE_SET = new SourceSet(HashSet.empty(), HashSet.empty());

  private final Set<Path> ignoredSources;

  public SourceSetRewriter(Set<Path> ignoredSources) {
    this.ignoredSources = ignoredSources;
  }

  public SourceSet rewrite(SourceSet input) {
    if (input.sources().size() == 1) {
      var singleSource = Paths.get(input.sources().head());
      if (ignoredSources.exists(singleSource::endsWith)) {
        return EMPTY_SOURCE_SET;
      }
    }
    return input;
  }
}

File: server/src/main/java/org/jetbrains/bsp/bazel/server/bloop/SourceSetRewriter.kt
package org.jetbrains.bsp.bazel.server.bloop

import org.jetbrains.bsp.bazel.server.sync.model.SourceSet
import java.nio.file.Path
import java.nio.file.Paths

class SourceSetRewriter(private val ignoredSources: Set<Path>) {
    fun rewrite(input: SourceSet): SourceSet {
        if (input.sources.size == 1) {
            val singleSource: Path = Paths.get(input.sources.first())
            if (ignoredSources.any(singleSource::endsWith)) {
                return EMPTY_SOURCE_SET
            }
        }
        return input
    }

    companion object {
        private val EMPTY_SOURCE_SET = SourceSet(HashSet(), HashSet())
    }
}
